lib/queuePopulator/IngestionPopulator.js

```diff
@@ -0,0 +1,364 @@
+                    sourceObj, this.extConfigs.ingestion.zookeeperPath);
+                return undefined;
+            });
+        }
+    }
+
+    _subscribeToLogSourceDispatcher(zkEndpoint, bucketd, ingestionPath) {
+        this.raftIdDispatcher =
+            new ProvisionDispatcher({ connectionString: zkEndpoint });
+        this.raftIdDispatcher.subscribe((err, items) => {
+            if (err) {
+                this.log.error(
+                    'error when receiving log source provision list',
+                    { zkEndpoint, error: err });
+                return undefined;
+            }
+            if (items.length === 0) {
+                this.log.info('no log source provisioned, idling',
+                              { zkEndpoint });
+            }
+            const newRaftReaders = items.map(
+                token => {
+                    if (token === 'mongo') {
+                        return new MongoLogReader({
+                            zkClient: this.zkClient,
+                            kafkaConfig: this.kafkaConfig,
+                            zkConfig: this.zkConfig,
+                            mongoConfig: this.qpConfig.mongo,
+                            logger: this.log,
+                            extensions: this._extensions,
+                            metricsProducer: this._mProducer,
+                        });
+                    }
+                    return new IngestionReader({
+                        zkClient: this.zkClient,
+                        kafkaConfig: this.kafkaConfig,
+                        bucketdConfig: bucketd,
+                        raftId: token,
+                        logger: this.log,
+                        extensions: this._extensions,
+                        metricsProducer: this._mProducer,
+                        ingestionPath,
+                        qpConfig: this.qpConfig,
+                        s3Config: this.s3Config,
+                    });
+                });
+            newRaftReaders.forEach(reader =>
+                this.logReadersUpdate.push(reader));
+            return undefined;
+        });
+        this.log.info('waiting to be provisioned a log source',
+                      { zkEndpoint });
+    }
+
+    _setupZookeeper(done) {
+        const zookeeperUrl = this.zkConfig.connectionString;
+        this.log.info('opening zookeeper connection for persisting ' +
+                      'populator state',
+                      { zookeeperUrl });
+        this.zkClient = zookeeper.createClient(zookeeperUrl, {
+            autoCreateNamespace: this.zkConfig.autoCreateNamespace,
+        });
+        this.zkClient.connect();
+        this.zkClient.once('error', done);
+        this.zkClient.once('ready', () => {
+            // just in case there would be more 'error' events emitted
+            this.zkClient.removeAllListeners('error');
+            done();
+        });
+    }
+
+    _loadExtensions() {
```

Comment:
We probably don't need to load the extensions for the ingestion daemon, keep for later refactor.